07-13-16-language-plugins.md

```diff
@@ -0,0 +1,97 @@
+
+```.js
+// This must implement a Promise that resolves with a Buffer
+// that represents a DIRECTORY containing the desired bundle.
+exportables.tarBundle = function(options) {
+  // Should return a Promise which eventually gets resolved with the bundle
+  return Promise.resolve(bundle);
+};
+```
+
+### 4. preRun
+`preRun` is an optional step that's useful if you need to call any shell functions on Tessel itself before the project is initialized. For example, with Rust, we set the permissions of the binary to ensure it's executable. It's unused for JavaScript.
+
+```.js
+// Optionally make changes within Tessel's Linux environment before starting the program
+exportables.preRun = function(tessel, options) {
+  // Should return a Promise which eventually gets resolved
+  return Promise.resolve();
+};
+```
+
+### 5. shell/binary
+#### t2 run
+`binary` is a string that represents the engine that runs the program. It is used as the first argument in a shell command to start the program when a project is *`run`*. For JavaScript, this string is `node` and for Rust it's `.`.
+
+
+#### t2 push
+`shell` is a function that could have been named `begin` because it's responsible for actually getting the program going when you *`push`* the project. The reason it's called `shell` instead is because the function needs to return a string representing the contents of a shell script that starts the program.
+
+```
+exportables.binary = 'YOUR_EXECUTION_METHOD';
+exportables.shell = (options) => {
+  return tags.stripIndent `
+    #!/bin/sh
+    exec YOUR_EXECUTION_METHOD /app/remote-script/${opts.resolvedEntryPoint}
+  `;
+};
+```
+
+### 6. postRun
+`postRun` is another optional step that's useful if there is code you need to run on Tessel after the program has started (but not necessarily before the program has ended). It's largely unused thus far so far but was added for future-proofness.
+
+```.js
+// Optionally make changes on Tessel after starting the program
+exportables.postRun = function(tessel, options) {
+  // Should return a Promise which eventually gets resolved
+  return Promise.resolve();
+};
+```
+
+### Conclusion
+That's just about there is to adding language support to the CLI. There are several more steps that need to be taken to build full hardware support for a language on Tessel and you can follow [Rust's progress in that process here](https://gist.github.com/johnnyman727/f76616edc71a8646b27b27809ea464c6). If you have questions about the unit tests or the implementation of new languages in general, feel free to reach out in the #engineering channel of [Tessel Slack](http://www.tessel-slack.herokuapp.com). Let us know if there are any extra plugin functions you think the CLI needs to accomodate or if you need help building a plugin for the language you'd like to build hardware with.
```

Comment:
Break up this sentence: "There are several more steps ... on Tessel. You can follow..."